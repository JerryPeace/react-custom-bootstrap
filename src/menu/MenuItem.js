import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classNames from 'classnames';
import styleMaps from '../styleMaps';
import faUtils from '../utils/faUtils';

class MenuItem extends Component {
    static propTypes = {
        active: PropTypes.bool,
        divider: PropTypes.bool,
        disabled: PropTypes.bool,
        deselectable: PropTypes.bool,
        faProps: PropTypes.object,
        header: PropTypes.bool,
        clickableHeader: PropTypes.bool,
        href: PropTypes.string,
        iconEl: PropTypes.node,
        noRootClose: PropTypes.bool,
        selectData: PropTypes.any,
        subMenu: PropTypes.node,
        onActiveClicked: PropTypes.func, // Dispatched when this menu item is active, not
                                         // deseletable, and clicked by user.
        onClick: PropTypes.func, // Dispatched when this menu is not disabled and clicked
        onItemSelect: PropTypes.func, // Dispatched when this menu is not disabled and clicked
        onSelect: PropTypes.func, // Dispatched when this menu is not disabled and clicked
        faStyle: PropTypes.string,
        sidemenu: PropTypes.node
    };

    static defaultProps = {
        divider: false,
        disabled: false,
        deselectable: false,
        header: false,
        noRootClose: false,
        href: ''
    };

    constructor(props, context) {
        super(props, context);

        this.handleClick = this.handleClick.bind(this);
    }

    handleClick(event) {
        const {
            children,
            href,
            deselectable,
            disabled,
            noRootClose,
            onActiveClicked,
            onItemSelect, // props from Menu
            onSelect,
            onClick,
            selectData,
            active
        } = this.props;

        if (!href || disabled) {
            event.preventDefault();
        }

        if (disabled || (active && !deselectable)) {
            active && onActiveClicked && onActiveClicked();
            return;
        }

        onSelect && onSelect({
            selectData,
            event,
            itemContent: children
        });

        onClick && onClick(event);

        onItemSelect && onItemSelect({
            selectData,
            event,
            itemContent: children
        });

        noRootClose && event.nativeEvent.stopImmediatePropagation();
    }

    renderIcon() {
        const { iconEl, faProps } = this.props;
        const faClass = faUtils.getCls(this.props);

        let icon = null;

        if (iconEl) {
            icon = iconEl;
        } else if (faClass) {
            icon = <i {...faProps} className={faClass} />;
        }

        return icon ? <span className='item-icon'>{icon}</span> : null;
    }

    renderDivider() {
        const { className } = this.props;
        const { divider } = styleMaps.CLASSES;

        return (
            <li className={classNames(className, divider)} />
        );
    }

    renderHeader() {
        const { children, className, clickableHeader } = this.props;
        const { 'dropdown-header': dropdownHeader } = styleMaps.CLASSES;

        return (
            <li className={classNames(className, dropdownHeader)}>
                {children}
            </li>
        );
    }

    renderClickableHeader() {
        const { children, className, clickableHeader } = this.props;
        const { 'dropdown-header': dropdownHeader } = styleMaps.CLASSES;

        return (
            <li
                className={classNames(className, dropdownHeader, 'clickable')}
                onClick={this.handleClick}
            >
                {children}
            </li>
        );
    }

    renderItem() {
        const {
            active,
            children,
            className,
            deselectable,
            disabled,
            divider,
            faProps,
            faStyle, // We don't use this props at this function but for extracting it from rest.
            header,
            clickableHeader,
            href,
            iconEl,
            noRootClose,
            onActiveClicked, // We don't use this props at this function but for extracting it from
                             // rest.
            onItemSelect,
            onSelect,
            selectData,
            sidemenu, // We don't use this props at this function but for extracting it from rest.
            subMenu,
            ...rest
        } = this.props;
        let style = null;

        if (subMenu) {
            style = {
                position: 'relative'
            };
        }

        return (
            <li
                {...rest}
                className={classNames(className, {
                    active,
                    disabled
                })}
                style={style}
                >
                <a
                    {...rest}
                    href={href}
                    onClick={this.handleClick}
                    >
                    {this.renderIcon()}
                    {children}
                </a>
                {subMenu && React.cloneElement(subMenu, {subMenu: true})}
            </li>
        );
    }

    render() {
        const { divider, header, clickableHeader } = this.props;

        if (divider) {
            return this.renderDivider();
        } else if (header) {
            return this.renderHeader();
        } else if (clickableHeader) {
            return this.renderClickableHeader();
        } else {
            return this.renderItem();
        }
    }
}

export default MenuItem;
