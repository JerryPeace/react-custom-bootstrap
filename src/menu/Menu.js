import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classNames from 'classnames';
import { bootstrapUtils } from '../utils';
import styleMaps from '../styleMaps';
import RootCloseWrapper from './RootCloseWrapper';
import MenuItem from './MenuItem';

class Menu extends Component {
    static propTypes = {
        bsClass: PropTypes.string,
        bsSize: PropTypes.string,
        bsStyle: PropTypes.string,
        isMenuOpen: PropTypes.bool,
        isMenuAlignRight: PropTypes.bool,
        isMenuCheckStyle: PropTypes.bool,
        isSubMenuOpenLeft: PropTypes.bool,
        noRootClose: PropTypes.bool,
        onRootClose: PropTypes.func,
        onItemClick: PropTypes.func,
        onItemSelect: PropTypes.func,
        sideMenu: PropTypes.element,
        style: PropTypes.object,
        subMenu: PropTypes.bool
    };

    static defaultProps = {
        bsClass: 'dropdown-menu',
        isMenuOpen: false,
        isMenuAlignRight: false,
        isMenuCheckStyle: false,
        isSubMenuOpenLeft: false,
        noRootClose: false
    };

    constructor(props) {
        super(props);
        this.state = {
            subMenuLeft: '100%' // open right side by default
        };
    }

    componentDidUpdate(prevProps) {
        const {
            noRootClose,
            subMenu,
            isMenuOpen,
            isSubMenuOpenLeft
        } = this.props;

        if (isMenuOpen
            && !prevProps.isMenuOpen
            && subMenu
            && isSubMenuOpenLeft
            && !noRootClose) {
            /* eslint-disable react/no-did-update-set-state */
            this.setState({ subMenuLeft: -this.node.getNode().getBoundingClientRect().width });
            /* eslint-enable react/no-did-update-set-state */
        }
    }

    render() {
        const {
            className,
            children,
            style,
            isMenuOpen,
            isMenuAlignRight,
            isMenuCheckStyle,
            subMenu,
            sideMenu,
            noRootClose,
            onRootClose,
            onItemSelect,
            isSubMenuOpenLeft,
            bsClass,
            bsSize,
            bsStyle,
            ...props
        } = this.props;
        const { subMenuLeft } = this.state;
        const bsCls = bootstrapUtils.getCls(this.props);
        let component = null;
        let menuStyle = {
            display: isMenuOpen ? 'block' : 'none'
        };

        if (subMenu) {
            menuStyle = {
                ...menuStyle,
                position: 'absolute',
                top: 0,
                left: subMenuLeft,
                marginTop: 0
            };
        } else if (sideMenu) {
            menuStyle = {
                ...menuStyle,
                display: isMenuOpen ? 'table-cell' : 'none',
                border: 'none',
                margin: 0,
                padding: 0,
                position: 'relative',
                top: 0,
                height: '100%',
                boxShadow: 'none',
                verticalAlign: 'top',
                float: 'none'
            };
        }

        component = (
            <ul
                {...props}
                className={classNames(className, bsCls, {
                    [styleMaps.CLASSES['check-style']]: isMenuCheckStyle,
                    [styleMaps.CLASSES['dropdown-menu-right']]: !sideMenu && isMenuAlignRight
                })}
                style={{
                    ...menuStyle,
                    ...style
                }}
            >
                {React.Children.map(children, (child) => {
                    if (child && child.type === MenuItem) {
                        return React.cloneElement(child, {onItemSelect});
                    } else {
                        return child;
                    }
                })}
            </ul>
        );

        if (!sideMenu && !noRootClose) {
            component = (
                <RootCloseWrapper
                    disabled={!isMenuOpen}
                    onRootClose={onRootClose}
                    ref={node => this.node = node}
                >
                    {component}
                </RootCloseWrapper>
            );
        }

        return component;
    }
}

export default Menu;
