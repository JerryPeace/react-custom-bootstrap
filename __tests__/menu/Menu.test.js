import React from 'react';
import { mount } from 'enzyme';
import Menu from 'menu/Menu';
import MenuItem from 'menu/MenuItem';

describe('Menu', () => {
    test('render Menu correctly', () => {
        const menu = mount(
            <Menu>
                <MenuItem header={true}>Menu Header</MenuItem>
                <MenuItem selectData='A'>Menu Enabled</MenuItem>
                <MenuItem divider={true}>Menu Disabled</MenuItem>
                <MenuItem selectData='B' disabled={true}>Menu Disabled</MenuItem>
                <div>simple component</div>
            </Menu>
        );
        expect(menu).toMatchSnapshot();
    });

    test('render Menu - sideMenu correctly', () => {
        const menu = mount(
            <Menu sideMenu={true}>
                <MenuItem header={true}>Menu Header</MenuItem>
                <MenuItem selectData='A'>Menu Enabled</MenuItem>
                <MenuItem divider={true}>Menu Disabled</MenuItem>
                <MenuItem selectData='B' disabled={true}>Menu Disabled</MenuItem>
                <div>simple component</div>
            </Menu>
        );
        expect(menu).toMatchSnapshot();
    });

    test('render Menu - subMenu correctly', () => {
        const menu = mount(
            <Menu subMenu={true}>
                <MenuItem header={true}>Menu Header</MenuItem>
                <MenuItem selectData='A'>Menu Enabled</MenuItem>
                <MenuItem divider={true}>Menu Disabled</MenuItem>
                <MenuItem selectData='B' disabled={true}>Menu Disabled</MenuItem>
                <div>simple component</div>
            </Menu>
        );
        expect(menu).toMatchSnapshot();
    });
});
