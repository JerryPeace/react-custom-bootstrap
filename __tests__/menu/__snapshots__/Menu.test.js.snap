// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu render Menu - sideMenu correctly 1`] = `
<Menu
  bsClass="dropdown-menu"
  isMenuAlignRight={false}
  isMenuCheckStyle={false}
  isMenuOpen={false}
  isSubMenuOpenLeft={false}
  noRootClose={false}
  sideMenu={true}
>
  <ul
    className="dropdown-menu"
    style={
      Object {
        "border": "none",
        "boxShadow": "none",
        "display": "none",
        "float": "none",
        "height": "100%",
        "margin": 0,
        "padding": 0,
        "position": "relative",
        "top": 0,
        "verticalAlign": "top",
      }
    }
  >
    <MenuItem
      deselectable={false}
      disabled={false}
      divider={false}
      header={true}
      href=""
      noRootClose={false}
    >
      <li
        className="dropdown-header"
      >
        Menu Header
      </li>
    </MenuItem>
    <MenuItem
      deselectable={false}
      disabled={false}
      divider={false}
      header={false}
      href=""
      noRootClose={false}
      selectData="A"
    >
      <li
        className=""
        style={null}
      >
        <a
          href=""
          onClick={[Function]}
        >
          Menu Enabled
        </a>
      </li>
    </MenuItem>
    <MenuItem
      deselectable={false}
      disabled={false}
      divider={true}
      header={false}
      href=""
      noRootClose={false}
    >
      <li
        className="divider"
      />
    </MenuItem>
    <MenuItem
      deselectable={false}
      disabled={true}
      divider={false}
      header={false}
      href=""
      noRootClose={false}
      selectData="B"
    >
      <li
        className="disabled"
        style={null}
      >
        <a
          href=""
          onClick={[Function]}
        >
          Menu Disabled
        </a>
      </li>
    </MenuItem>
    <div>
      simple component
    </div>
  </ul>
</Menu>
`;

exports[`Menu render Menu - subMenu correctly 1`] = `
<Menu
  bsClass="dropdown-menu"
  isMenuAlignRight={false}
  isMenuCheckStyle={false}
  isMenuOpen={false}
  isSubMenuOpenLeft={false}
  noRootClose={false}
  subMenu={true}
>
  <RootCloseWrapper
    disabled={true}
    event="click"
  >
    <ul
      className="dropdown-menu"
      style={
        Object {
          "display": "none",
          "left": "100%",
          "marginTop": 0,
          "position": "absolute",
          "top": 0,
        }
      }
    >
      <MenuItem
        deselectable={false}
        disabled={false}
        divider={false}
        header={true}
        href=""
        noRootClose={false}
      >
        <li
          className="dropdown-header"
        >
          Menu Header
        </li>
      </MenuItem>
      <MenuItem
        deselectable={false}
        disabled={false}
        divider={false}
        header={false}
        href=""
        noRootClose={false}
        selectData="A"
      >
        <li
          className=""
          style={null}
        >
          <a
            href=""
            onClick={[Function]}
          >
            Menu Enabled
          </a>
        </li>
      </MenuItem>
      <MenuItem
        deselectable={false}
        disabled={false}
        divider={true}
        header={false}
        href=""
        noRootClose={false}
      >
        <li
          className="divider"
        />
      </MenuItem>
      <MenuItem
        deselectable={false}
        disabled={true}
        divider={false}
        header={false}
        href=""
        noRootClose={false}
        selectData="B"
      >
        <li
          className="disabled"
          style={null}
        >
          <a
            href=""
            onClick={[Function]}
          >
            Menu Disabled
          </a>
        </li>
      </MenuItem>
      <div>
        simple component
      </div>
    </ul>
  </RootCloseWrapper>
</Menu>
`;

exports[`Menu render Menu correctly 1`] = `
<Menu
  bsClass="dropdown-menu"
  isMenuAlignRight={false}
  isMenuCheckStyle={false}
  isMenuOpen={false}
  isSubMenuOpenLeft={false}
  noRootClose={false}
>
  <RootCloseWrapper
    disabled={true}
    event="click"
  >
    <ul
      className="dropdown-menu"
      style={
        Object {
          "display": "none",
        }
      }
    >
      <MenuItem
        deselectable={false}
        disabled={false}
        divider={false}
        header={true}
        href=""
        noRootClose={false}
      >
        <li
          className="dropdown-header"
        >
          Menu Header
        </li>
      </MenuItem>
      <MenuItem
        deselectable={false}
        disabled={false}
        divider={false}
        header={false}
        href=""
        noRootClose={false}
        selectData="A"
      >
        <li
          className=""
          style={null}
        >
          <a
            href=""
            onClick={[Function]}
          >
            Menu Enabled
          </a>
        </li>
      </MenuItem>
      <MenuItem
        deselectable={false}
        disabled={false}
        divider={true}
        header={false}
        href=""
        noRootClose={false}
      >
        <li
          className="divider"
        />
      </MenuItem>
      <MenuItem
        deselectable={false}
        disabled={true}
        divider={false}
        header={false}
        href=""
        noRootClose={false}
        selectData="B"
      >
        <li
          className="disabled"
          style={null}
        >
          <a
            href=""
            onClick={[Function]}
          >
            Menu Disabled
          </a>
        </li>
      </MenuItem>
      <div>
        simple component
      </div>
    </ul>
  </RootCloseWrapper>
</Menu>
`;
